File: matrix_mpi.py
import os
# numpy ın kendi kendine paralelleşmesini yasaklama muhabbeti
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"] = "1"

from mpi4py import MPI
import numpy as np
import json

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

N = 0
A = None
B = None

if rank == 0:
    with open('a.txt', 'r') as f:
        N = int(f.readline().strip())
        A = np.loadtxt(f, dtype=np.float64)
        
    with open('b.txt', 'r') as f:
        f.readline()
        B = np.loadtxt(f, dtype=np.float64)
        
    A = np.ascontiguousarray(A, dtype=np.float64) 
    start_total = MPI.Wtime()
else:
    B = None

N = comm.bcast(N, root=0)
if rank != 0:
    B = np.empty((N, N), dtype=np.float64)
comm.Bcast(B, root=0)

satir_sayisi = N // size
local_A = np.empty((satir_sayisi, N), dtype=np.float64)
comm.Scatter(A, local_A, root=0)

start_comp = MPI.Wtime()
local_C = np.dot(local_A, B) 
end_comp = MPI.Wtime()
t_comp = end_comp - start_comp

if rank == 0:
    C = np.empty((N, N), dtype=np.float64)
else:
    C = None
comm.Gather(local_C, C, root=0)

// json kaydewtme burası
if rank == 0:
    end_total = MPI.Wtime()
    t_total = end_total - start_total
    
    cpu_modeli = "i5-10400" 
    dil = "Python"
    dosya_adi = "sonuclar_10400.json"
    veriler = []
    
    if os.path.exists(dosya_adi):
        with open(dosya_adi, "r") as f:
            try:
                veriler = json.load(f)
            except json.JSONDecodeError:
                veriler = [] 
                
    kayit_bulundu = False
    for kayit in veriler:
        if kayit.get("CPU") == cpu_modeli and kayit.get("Dil") == dil and kayit.get("P") == size and kayit.get("N") == N:
            kayit["T_total"] = t_total
            kayit["T_comp"] = t_comp
            kayit_bulundu = True
            break
            
    if not kayit_bulundu:
        yeni_veri = {
            "CPU": cpu_modeli,
            "Dil": dil,
            "N": N,
            "P": size,
            "T_total": t_total,
            "T_comp": t_comp
        }
        veriler.append(yeni_veri)
    
    with open(dosya_adi, "w") as f:
        json.dump(veriler, f, indent=4)
        
    print(f"P={size} test edildi. Toplam Sure: {t_total:.6f} sn | Saf Hesaplama: {t_comp:.6f} sn")

File: matrix_mpi.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <cblas.h> 

int main(int argc, char** argv) {
    int rank, size, N;
    double *A = NULL, *B = NULL, *C = NULL;
    double *local_A, *local_C;
    double start_total, end_total, start_comp, end_comp;
    double t_total, t_comp;

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);


    if (rank == 0) {
        FILE *fa = fopen("a.txt", "r");
        fscanf(fa, "%d", &N); 
        
        A = (double*)malloc(N * N * sizeof(double));
        for(int i = 0; i < N * N; i++) {
            fscanf(fa, "%lf", &A[i]); // A matrisini doldur
        }
        fclose(fa);

        FILE *fb = fopen("b.txt", "r");
        fscanf(fb, "%*d"); // b.txt'nin boyut satırını atla
        B = (double*)malloc(N * N * sizeof(double));
        for(int i = 0; i < N * N; i++) {
            fscanf(fb, "%lf", &B[i]); // B matrisini doldur
        }
        fclose(fb);

        start_total = MPI_Wtime(); 
    }

    
    MPI_Bcast(&N, 1, MPI_INT, 0, MPI_COMM_WORLD); 
    
    if (rank != 0) {
        B = (double*)malloc(N * N * sizeof(double)); 
    }
    MPI_Bcast(B, N * N, MPI_DOUBLE, 0, MPI_COMM_WORLD); 

    int satir_sayisi = N / size;
    int eleman_sayisi = satir_sayisi * N;

    local_A = (double*)malloc(eleman_sayisi * sizeof(double));
    local_C = (double*)malloc(eleman_sayisi * sizeof(double));

    MPI_Scatter(A, eleman_sayisi, MPI_DOUBLE, local_A, eleman_sayisi, MPI_DOUBLE, 0, MPI_COMM_WORLD);


    start_comp = MPI_Wtime();
    
    // cblas_dgemm: C = alpha * A * B + beta * C
    cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, 
                satir_sayisi, N, N, 
                1.0, local_A, N, B, N, 
                0.0, local_C, N);
                
    end_comp = MPI_Wtime();
    t_comp = end_comp - start_comp;


    if (rank == 0) {
        C = (double*)malloc(N * N * sizeof(double));
    }

    MPI_Gather(local_C, eleman_sayisi, MPI_DOUBLE, C, eleman_sayisi, MPI_DOUBLE, 0, MPI_COMM_WORLD);

    
    if (rank == 0) { // bundan sonrası json kaydetme muhabbeti
        end_total = MPI_Wtime();
        t_total = end_total - start_total;

        printf("Dil: C | N=%d | P=%d test edildi. Toplam Sure: %f sn | Saf Hesaplama: %f sn\n", N, size, t_total, t_comp);

        char cmd[256];
        snprintf(cmd, sizeof(cmd), "python3 update_json.py i5-10400 C %d %d %f %f", N, size, t_total, t_comp);
        system(cmd);

        free(A); free(C);
    }

    free(B); free(local_A); free(local_C);
    MPI_Finalize();
    return 0;
}